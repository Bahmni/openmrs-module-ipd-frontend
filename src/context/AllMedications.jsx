import React, { useState } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { ALL_DRUG_ORDERS_URL } from "../constants";

export const AllMedicationsContext = React.createContext({
  getAllDrugOrders: () => {},
});

export const AllMedicationsContextProvider = (props) => {
  const getAllDrugOrders = async (visitUuid) => {
    setMedications((prevState) => ({ ...prevState, isLoading: true }));
    try {
      const response = await axios.get(ALL_DRUG_ORDERS_URL(visitUuid), {
        withCredentials: true,
      });
      if (response.status !== 200) throw new Error(response.statusText);
      setMedications((prevState) => ({
        ...prevState,
        isLoading: false,
        data: response.data,
      }));
    } catch (error) {
      setMedications((prevState) => ({
        ...prevState,
        isLoading: false,
        error,
      }));
    }
  };

  const initialState = {
    isLoading: false,
    data: null,
    error: null,
    getAllDrugOrders,
  };

  const [medications, setMedications] = useState(initialState);

  return (
    <AllMedicationsContext.Provider value={medications}>
      {props.children}
    </AllMedicationsContext.Provider>
  );
};

AllMedicationsContextProvider.propTypes = {
  children: PropTypes.children,
};
