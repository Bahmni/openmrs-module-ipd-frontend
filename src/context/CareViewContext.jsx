import React from "react";

export const CareViewContext = React.createContext();
