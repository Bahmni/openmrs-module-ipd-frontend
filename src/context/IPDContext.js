import React from "react";

export const IPDContext = React.createContext();