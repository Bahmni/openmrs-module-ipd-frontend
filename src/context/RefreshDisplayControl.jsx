import React from "react";

const RefreshDisplayControl = React.createContext({
  refreshDisplayControl: () => {},
});

export default RefreshDisplayControl;
