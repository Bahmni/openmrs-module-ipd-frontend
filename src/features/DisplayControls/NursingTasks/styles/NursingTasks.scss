.no-nursing-tasks {
  padding-top: 10px;
}

.nursing-tasks-content-container {
  display: flex;
  flex-direction: column;
  position: relative;

  .bx--btn--expressive {
    font-size: 14px;
    &:hover,
    &:focus {
      color: #fff;
    }
  }

  .nursing-task-navigation {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin: 10px 0;
  }

  .nursing-task-actions {
    display: flex;
    gap: 10px;
  }

  .nursing-task-dropdown {
    min-width: 130px;
  }
}

.nursing-task-tiles-container {
  display: flex;
  flex-wrap: wrap;
}

.date-text {
  font-size: 14px;
  font-style: normal;
  font-weight: 400;
  line-height: 14px;
  letter-spacing: 0.14px;
}
