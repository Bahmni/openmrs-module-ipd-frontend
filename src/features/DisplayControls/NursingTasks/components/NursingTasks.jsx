import React, { useContext, useEffect, useState } from "react";
import PropTypes from "prop-types";
import { FormattedMessage } from "react-intl";
import "../styles/NursingTasks.scss";

import {
  fetchMedicationNursingTasks,
  ExtractMedicationNursingTasksData,
  GetUTCEpochForDate,
} from "../utils/NursingTasksUtils";
import TaskTile from "./TaskTile";
import { formatDate } from "../../../../utils/DateTimeUtils";
import { SliderContext } from "../../../../context/SliderContext";
import UpdateNursingTasks from "./UpdateNursingTasks";

export default function NursingTasks(props) {
  const { patientId } = props;
  const [medicationNursingTasks, setMedicationNursingTasks] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const { isSliderOpen, updateSliderOpen } = useContext(SliderContext);
  const [selectedMedicationTask, setSelectedMedicationTask] = useState([]);
  const updateNursingTasksSlider = (value) => {
    updateSliderOpen((prev) => {
      return {
        ...prev,
        nursingTasks: value,
      };
    });
  };
  const NoNursingTasksMessage = (
    <FormattedMessage
      id={"NO_NURSING_TASKS_MESSAGE"}
      defaultMessage={"No nursing task is scheduled for the patient"}
    />
  );

  const fetchNursingTasks = async () => {
    setIsLoading(true);
    const forDate = GetUTCEpochForDate(new Date().toUTCString());
    const nursingTasks = await fetchMedicationNursingTasks(patientId, forDate);
    if (nursingTasks) {
      const extractedData = ExtractMedicationNursingTasksData(nursingTasks);
      setMedicationNursingTasks(extractedData);
      setIsLoading(false);
    }
  };
  useEffect(() => {
    fetchNursingTasks();
  }, []);

  const showCurrentDate = () => {
    var currentDate = new Date();

    var formattedDate = formatDate(currentDate, "DD/MM/YYYY");
    return <div className="date-text">{formattedDate}</div>;
  };

  const showTaskTiles = () => {
    return medicationNursingTasks.map((medicationNursingTask, index) => {
      return (
        <div key={index}>
          <div
            onClick={() => {
              if (!isSliderOpen.nursingTasks) {
                setSelectedMedicationTask(medicationNursingTask);
                updateNursingTasksSlider(true);
              }
            }}
          >
            <TaskTile medicationNursingTask={medicationNursingTask} />
          </div>
        </div>
      );
    });
  };

  const showMedicationNursingTasks = () => {
    if (isLoading) {
      return <div>Loading...</div>;
    }
    if (medicationNursingTasks && medicationNursingTasks.length === 0) {
      return <div className="no-nursing-tasks">{NoNursingTasksMessage}</div>;
    }

    return (
      <div className="nursing-tasks-content-container">
        {showCurrentDate()}
        {isSliderOpen.nursingTasks && (
          <UpdateNursingTasks
            medicationTasks={selectedMedicationTask}
            updateNursingTasksSlider={updateNursingTasksSlider}
          />
        )}
        <div className="nursing-task-tiles-container">{showTaskTiles()}</div>
      </div>
    );
  };
  return <div>{showMedicationNursingTasks()}</div>;
}

NursingTasks.propTypes = {
  patientId: PropTypes.string.isRequired,
};
