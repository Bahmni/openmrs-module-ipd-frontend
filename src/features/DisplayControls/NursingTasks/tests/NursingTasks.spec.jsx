import React from "react";
import { render, waitFor } from "@testing-library/react";
import NursingTasks from "../components/NursingTasks";
import { mockNursingTasksResponse } from "./NursingTasksUtilsMockData";
import MockDate from "mockdate";
import { SliderContext } from "../../../../context/SliderContext";

const mockFetchMedicationNursingTasks = jest.fn();

jest.mock("../utils/NursingTasksUtils", () => {
  const originalModule = jest.requireActual("../utils/NursingTasksUtils");
  return {
    ...originalModule,
    fetchMedicationNursingTasks: () => mockFetchMedicationNursingTasks(),
  };
});

const mockProviderValue = {
  isSliderOpen: {
    treatments: false,
  },
  updateSliderOpen: jest.fn(),
};
describe("NursingTasks", () => {
  afterEach(() => {
    jest.resetAllMocks();
    MockDate.reset();
  });

  it("should show loading state", async () => {
    const { getByText } = render(
      <SliderContext.Provider value={mockProviderValue}>
        <NursingTasks patientId="patientid" />
      </SliderContext.Provider>
    );

    await waitFor(() => {
      expect(getByText("Loading...")).toBeTruthy();
    });
  });

  it("should show no nursing tasks message when nursing task reponse is empty", async () => {
    mockFetchMedicationNursingTasks.mockResolvedValueOnce([]);
    const { getByText } = render(
      <SliderContext.Provider value={mockProviderValue}>
        <NursingTasks patientId="patientid" />
      </SliderContext.Provider>
    );

    await waitFor(() => {
      expect(mockFetchMedicationNursingTasks).toHaveBeenCalledTimes(1);
      expect(
        getByText("No Nursing Task is scheduled for the patient")
      ).toBeTruthy();
    });
  });

  it("should show nursing tasks when nursing task reponse is not empty", async () => {
    MockDate.set("2023-08-11");
    mockFetchMedicationNursingTasks.mockResolvedValueOnce(
      mockNursingTasksResponse
    );
    const { getAllByText, asFragment } = render(
      <SliderContext.Provider value={mockProviderValue}>
        <NursingTasks patientId="patientid" />
      </SliderContext.Provider>
    );
    await waitFor(() => {
      expect(mockFetchMedicationNursingTasks).toHaveBeenCalledTimes(1);
      expect(
        getAllByText("Paracetamol 120 mg/5 mL Suspension (Liquid)")
      ).toBeTruthy();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("should show current date", async () => {
    MockDate.set("2023-08-11");

    mockFetchMedicationNursingTasks.mockResolvedValueOnce(
      mockNursingTasksResponse
    );
    const { getByText } = render(
      <SliderContext.Provider value={mockProviderValue}>
        <NursingTasks patientId="patientid" />
      </SliderContext.Provider>
    );
    await waitFor(() => {
      expect(mockFetchMedicationNursingTasks).toHaveBeenCalledTimes(1);
      expect(getByText("11/08/2023")).toBeTruthy();
    });
  });
});
