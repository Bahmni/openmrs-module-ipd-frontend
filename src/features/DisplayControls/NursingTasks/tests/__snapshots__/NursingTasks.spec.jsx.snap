// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NursingTasks should show nursing tasks when nursing task reponse is not empty 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nursing-tasks-content-container"
    >
      <div
        class="date-text"
      >
        11/08/2023
      </div>
      <div
        class="nursing-task-tiles-container"
      >
        <div>
          <div
            class="tile-parent-container"
          >
            <div
              class="nursing-tasks-tile"
              style="background-color: rgb(237, 248, 230);"
            >
              <div
                class="tile-content"
              >
                <div
                  class="tile-title"
                >
                  <div
                    class="nursing-task-icon-container"
                  >
                    <div>
                      <test-file-stub />
                    </div>
                  </div>
                  <span
                    class="bx--tooltip--definition bx--tooltip--a11y"
                  >
                    <button
                      aria-describedby="definition-tooltip-1"
                      class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition bx--tooltip--bottom bx--tooltip--align-start"
                      type="button"
                    >
                      <div
                        class="drug-title"
                        style="color: rgb(169, 169, 169);"
                      >
                        Paracetamol 120 mg/5 mL Suspension (Liquid)
                      </div>
                    </button>
                    <span
                      class="bx--assistive-text"
                      id="definition-tooltip-1"
                      role="tooltip"
                    >
                      Paracetamol 120 mg/5 mL Suspension (Liquid)
                    </span>
                  </span>
                </div>
                <div
                  class="tile-content-subtext"
                  style="color: rgb(169, 169, 169);"
                >
                  <span>
                    25
                  </span>
                  <span>
                     - mg
                  </span>
                  <span>
                     - Oral
                  </span>
                </div>
                <div
                  class="tile-content-subtext"
                >
                  <test-file-stub />
                  <div>
                     04:15 PM
                  </div>
                </div>
                <div
                  class="more-info"
                >
                  (1 more)
                </div>
              </div>
            </div>
            <div
              class="nursing-tasks-tile stacked-tile"
              style="background-color: rgb(237, 248, 230);"
            />
          </div>
        </div>
        <div>
          <div
            class="tile-parent-container"
          >
            <div
              class="nursing-tasks-tile"
              style="background-color: rgb(237, 248, 230);"
            >
              <div
                class="tile-content"
              >
                <div
                  class="tile-title"
                >
                  <div
                    class="nursing-task-icon-container"
                  >
                    <div>
                      <test-file-stub />
                    </div>
                  </div>
                  <span
                    class="bx--tooltip--definition bx--tooltip--a11y"
                  >
                    <button
                      aria-describedby="definition-tooltip-2"
                      class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition bx--tooltip--bottom bx--tooltip--align-start"
                      type="button"
                    >
                      <div
                        class="drug-title"
                        style="color: rgb(169, 169, 169);"
                      >
                        Paracetamol 120 mg (Tablet)
                      </div>
                    </button>
                    <span
                      class="bx--assistive-text"
                      id="definition-tooltip-2"
                      role="tooltip"
                    >
                      Paracetamol 120 mg (Tablet)
                    </span>
                  </span>
                </div>
                <div
                  class="tile-content-subtext"
                  style="color: rgb(169, 169, 169);"
                >
                  <span>
                    25
                  </span>
                  <span>
                     - mg
                  </span>
                  <span>
                     - Oral
                  </span>
                </div>
                <div
                  class="tile-content-subtext"
                >
                  <test-file-stub />
                  <div>
                     04:32 PM
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
